### PR TITLE
Add expirationWatcher

### PR DESCRIPTION
**Pull request by LogvinovLeon**
This PR:
* Adds ExpirationWatcher which is used inside of OrderStateWatcher to report expired orders
* Adds tests for it

### CODE REVIEW

File: packages/0x.js/src/order_watcher/expiration_watcher.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        this.orderExpirationCheckingIntervalMs = orderExpirationCheckingIntervalMs ||
+                                                 DEFAULT_ORDER_EXPIRATION_CHECKING_INTERVAL_MS;
+        this.expiration = {};
+        const scoreFunction = ((orderHash: string) => {
+            return this.expiration[orderHash].toNumber();
+        }).bind(this);
+        this.orderHashHeapByExpiration = new Heap(scoreFunction);
+    }
+    public subscribe(callback: (orderHash: string) => void): void {
+        this.callbackIfExists = callback;
```
Comment (BMillman19) [inline packages/0x.js/src/order_watcher/expiration_watcher.ts]: if the callback already exists we should unsubscribe first

File: packages/0x.js/src/order_watcher/expiration_watcher.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        this.orderExpirationCheckingIntervalIdIfExists = intervalUtils.setAsyncExcludingInterval(
+            this.pruneExpiredOrders.bind(this), this.orderExpirationCheckingIntervalMs,
+        );
+    }
+    public unsubscribe(): void {
+        intervalUtils.clearAsyncExcludingInterval(this.orderExpirationCheckingIntervalIdIfExists as NodeJS.Timer);
```
Comment (BMillman19) [inline packages/0x.js/src/order_watcher/expiration_watcher.ts]: we should check if `this.orderExpirationCheckingIntervalIdIfExists` is not undefined before clearing

File: packages/0x.js/src/order_watcher/expiration_watcher.ts (hunk 1 above)
Comment (BMillman19) [inline packages/0x.js/src/order_watcher/expiration_watcher.ts]: we can take an approach similar to event_watcher here where we dont actually need to hold on to a reference to the callback we can just pass it thru to pruneExpiredOrders, that way we have less state to manage in this class

File: packages/0x.js/src/order_watcher/expiration_watcher.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        delete this.callbackIfExists;
```
Comment (BMillman19) [inline packages/0x.js/src/order_watcher/expiration_watcher.ts]: we should `delete this.orderExpirationCheckingIntervalIdIfExists` here

File: packages/0x.js/src/utils/heap.ts [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+            }
+
+            // Otherwise, swap the parent with the current element and
+            // continue.
+            this.content[parentN] = element;
+            this.content[n] = parent;
+            n = parentN;
+        }
+    }
+
```
Comment (BMillman19) [inline packages/0x.js/src/utils/heap.ts]: no new line between functions

File: packages/0x.js/src/order_watcher/expiration_watcher.ts [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        this.expiration[orderHash] = expirationUnixTimestampSec;
+        // We don't remove hashes on order remove because it's slow (linear).
+        // We just skip them later if the order was already removed from the order watcher.
+        this.orderHashHeapByExpiration.push(orderHash);
+    }
+    private pruneExpiredOrders(): void {
+        const currentUnixTimestampSec = utils.getCurrentUnixTimestamp();
+        while (
+            this.orderHashHeapByExpiration.size() !== 0 &&
+            this.expiration[this.orderHashHeapByExpiration.head()].lessThan(currentUnixTimestampSec) &&
```
Comment (abandeali1) [inline packages/0x.js/src/order_watcher/expiration_watcher.ts]: Realistically, people will probably want to prune orders more aggressively than this (probably orders that are within 15-30 seconds of expiration). You could use `this.expiration[this.orderHashHeapByExpiration.head()].lessThan(currentUnixTimestampSec + this.pruneOrderIfLessThanSec)`

File: packages/0x.js/src/order_watcher/order_state_watcher.ts [hunk 6]
```diff
@@ -84,6 +90,8 @@ export class OrderStateWatcher {
         assert.isValidSignature(orderHash, signedOrder.ecSignature, signedOrder.maker);
         this._orderByOrderHash[orderHash] = signedOrder;
         this.addToDependentOrderHashes(signedOrder, orderHash);
+        // We don't remove orders from expirationWatcher because heap removal is linear. We just skip it later
```
Comment (fabioberger) [inline packages/0x.js/src/order_watcher/order_state_watcher.ts]: Won't the heap grow into a huge memory leak if tons of orders with very length expirations are never removed from it?

File: packages/0x.js/src/order_watcher/order_state_watcher.ts [hunk 7]
```diff
@@ -123,6 +132,18 @@ export class OrderStateWatcher {
...
+    }
+    private _onOrderExpired(orderHash: string): void {
+        const orderState: OrderState = {
+            isValid: false,
+            orderHash,
+            error: ExchangeContractErrs.OrderFillExpired,
+        };
+        if (!_.isUndefined(this._orderByOrderHash[orderHash])) {
+            // We need this check because we never remove the orders from expiration watcher
+            (this._callbackIfExistsAsync as OnOrderStateChangeCallback)(orderState);
```
Comment (fabioberger) [inline packages/0x.js/src/order_watcher/order_state_watcher.ts]: We should probably check if `_callbackIfExistsAsync` is defined. Then we don't need to cast it.

File: packages/0x.js/src/types.ts [hunk 8]
```diff
@@ -397,9 +397,11 @@ export interface JSONRPCPayload {
 }
 
 /*
+ * orderExpirationCheckingIntervalMs: How often to check for expired orders
  * eventPollingIntervalMs: How often to poll the Ethereum node for new events
```
Comment (fabioberger) [inline packages/0x.js/src/types.ts]: I think for both of these we should add what the default values are set to. Thoughts?

File: packages/0x.js/test/expiration_watcher_test.ts [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+import {chaiSetup} from './utils/chai_setup';
+import {web3Factory} from './utils/web3_factory';
+import {utils} from '../src/utils/utils';
+import {Web3Wrapper} from '../src/web3_wrapper';
+import {TokenUtils} from './utils/token_utils';
+import {ExpirationWatcher} from '../src/order_watcher/expiration_watcher';
+import {
+    ZeroEx,
+    Token,
+} from '../src';
```
Comment (fabioberger) [inline packages/0x.js/test/expiration_watcher_test.ts]: Can we go back to referencing `types.ts` directly?

File: packages/0x.js/test/expiration_watcher_test.ts [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+        currentUnixTimestampSec = utils.getCurrentUnixTimestamp();
+        expirationWatcher = new ExpirationWatcher();
+    });
+    afterEach(() => {
+        timer.restore();
+        expirationWatcher.unsubscribe();
+    });
+    it('correctly emits events when order expires', (done: DoneCallback) => {
+        (async () => {
+            const orderLifetime = 60;
```
Comment (fabioberger) [inline packages/0x.js/test/expiration_watcher_test.ts]: Is seconds? Miliseconds?

File: packages/0x.js/test/expiration_watcher_test.ts [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+            const expirationUnixTimestampSec = currentUnixTimestampSec.plus(orderLifetime);
+            const signedOrder = await fillScenarios.createFillableSignedOrderAsync(
+                makerTokenAddress, takerTokenAddress, makerAddress, takerAddress, fillableAmount,
+                expirationUnixTimestampSec,
+            );
+            const orderHash = ZeroEx.getOrderHashHex(signedOrder);
+            expirationWatcher.addOrder(orderHash, signedOrder.expirationUnixTimestampSec);
+            const callback = reportCallbackErrors(done)((hash: string) => {
+                done(new Error('Emited expiration vent before the order actually expired'));
```
Comment (fabioberger) [inline packages/0x.js/test/expiration_watcher_test.ts]: typo: `emitted`

Comment (fabioberger) [reply packages/0x.js/test/expiration_watcher_test.ts]: typo: `vent`?

File: packages/0x.js/src/order_watcher/expiration_watcher.ts (hunk 1 above)
Comment (LogvinovLeon) [reply packages/0x.js/src/order_watcher/expiration_watcher.ts]: This is a private class that only we use, but still added an exception.

File: packages/0x.js/src/order_watcher/expiration_watcher.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (LogvinovLeon) [reply packages/0x.js/src/order_watcher/expiration_watcher.ts]: I can't cause that's within a loop

File: packages/0x.js/src/order_watcher/order_state_watcher.ts (hunk 6 above)
Comment (LogvinovLeon) [reply packages/0x.js/src/order_watcher/order_state_watcher.ts]: Will fix in a next PR

File: packages/0x.js/src/order_watcher/expiration_watcher.ts [hunk 12]
```diff
@@ -1,59 +1,69 @@
...
+        delete this.expiration[orderHash];
     }
-    private pruneExpiredOrders(): void {
-        const currentUnixTimestampSec = utils.getCurrentUnixTimestamp();
+    private pruneExpiredOrders(callback: (orderHash: string) => void): void {
+        const currentUnixTimestampMs = utils.getCurrentUnixTimestampMs();
         while (
-            this.orderHashHeapByExpiration.size() !== 0 &&
-            this.expiration[this.orderHashHeapByExpiration.head()].lessThan(currentUnixTimestampSec) &&
-            !_.isUndefined(this.callbackIfExists)
```
Comment (fabioberger) [inline packages/0x.js/src/order_watcher/expiration_watcher.ts]: Don't we want to replace this check with `!_.isUndefined(this.orderExpirationCheckingIntervalIdIfExists)`?

File: packages/0x.js/src/order_watcher/order_state_watcher.ts [hunk 13]
```diff
@@ -145,6 +151,7 @@ export class OrderStateWatcher {
         };
         if (!_.isUndefined(this._orderByOrderHash[orderHash])) {
             // We need this check because we never remove the orders from expiration watcher
```
Comment (fabioberger) [inline packages/0x.js/src/order_watcher/order_state_watcher.ts]: Outdated comment. Remove.

File: packages/0x.js/src/order_watcher/order_state_watcher.ts (hunk 7 above)
Comment (fabioberger) [reply packages/0x.js/src/order_watcher/order_state_watcher.ts]: @LogvinovLeon please address this feedback.

File: packages/0x.js/src/types.ts [hunk 14]
```diff
@@ -391,14 +391,18 @@ export interface JSONRPCPayload {
     method: string;
 }
 
+// tslint:disable:max-line-length
 /*
- * orderExpirationCheckingIntervalMs: How often to check for expired orders
- * eventPollingIntervalMs: How often to poll the Ethereum node for new events
+ * orderExpirationCheckingIntervalMs: How often to check for expired orders. Default: 50
+ * eventPollingIntervalMs: How often to poll the Ethereum node for new events. Defaults: 200
+ * expirationMarginMs: Amount of time before order expiry that you'd like to be notified of an orders expiration. Defaults: 0
```
Comment (fabioberger) [inline packages/0x.js/src/types.ts]: Let's just make this comment two lines and remove the `tslint:disable:max-line-length`

Comment (fabioberger) [reply packages/0x.js/src/types.ts]:
```
* expirationMarginMs: Amount of time before order expiry that you'd like to be notified 
*                     of an orders expiration. Defaults: 0
`

File: packages/0x.js/src/utils/utils.ts [hunk 15]
```diff
@@ -49,7 +49,10 @@ export const utils = {
         const hashHex = ethUtil.bufferToHex(hashBuff);
         return hashHex;
     },
-    getCurrentUnixTimestamp(): BigNumber {
-        return new BigNumber(Date.now() / 1000);
+    getCurrentUnixTimestampSec(): BigNumber {
+        return new BigNumber(Date.now() / 1000).round();
+    },
+    getCurrentUnixTimestampMs(): BigNumber {
+        return new BigNumber(Date.now());
```
Comment (fabioberger) [inline packages/0x.js/src/utils/utils.ts]: Nice!

File: packages/0x.js/test/expiration_watcher_test.ts [hunk 16]
```diff
@@ -10,10 +10,8 @@ import {utils} from '../src/utils/utils';
 import {Web3Wrapper} from '../src/web3_wrapper';
 import {TokenUtils} from './utils/token_utils';
 import {ExpirationWatcher} from '../src/order_watcher/expiration_watcher';
-import {
-    ZeroEx,
-    Token,
-} from '../src';
+import {Token} from '../src/types';
+import {ZeroEx} from '../src';
```
Comment (fabioberger) [inline packages/0x.js/test/expiration_watcher_test.ts]: `../src/0x`?

File: packages/0x.js/test/expiration_watcher_test.ts [hunk 17]
```diff
@@ -68,8 +66,8 @@ describe('ExpirationWatcher', () => {
     });
     it('correctly emits events when order expires', (done: DoneCallback) => {
         (async () => {
-            const orderLifetime = 60;
-            const expirationUnixTimestampSec = currentUnixTimestampSec.plus(orderLifetime);
+            const orderLifetimeS = 60;
```
Comment (fabioberger) [inline packages/0x.js/test/expiration_watcher_test.ts]: nit: `Sec` rather then `S`?

Comment (fabioberger) [reply packages/0x.js/test/expiration_watcher_test.ts]: Looks like a possible typo: `Lifetimes` is also an english word ;)

File: packages/0x.js/src/order_watcher/expiration_watcher.ts [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+        this.orderHashByExpirationRBTree.insert(orderHash);
+    }
+    public removeOrder(orderHash: string): void {
+        this.orderHashByExpirationRBTree.remove(orderHash);
+        delete this.expiration[orderHash];
+    }
+    private async pruneExpiredOrdersAsync(callbackAsync: (orderHash: string) => Promise<void>): Promise<void> {
+        const currentUnixTimestampMs = utils.getCurrentUnixTimestampMs();
+        while (true) {
+            const noOrdersLeft = this.orderHashByExpirationRBTree.size === 0;
```
Comment (fabioberger) [inline packages/0x.js/src/order_watcher/expiration_watcher.ts]: Nit: boolen variable names should start with `is`, `has`, `does`, `should`. Rename `noOrdersLeft` to `isTrackingOrders`

```
const isTrackingOrders = this.orderHashByExpirationRBTree.size !== 0;

if (!isTrackingOrders) {
```

File: packages/0x.js/src/order_watcher/expiration_watcher.ts [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+            if (noOrdersLeft) {
+                break;
+            }
+            const nextOrderHashToExpire = this.orderHashByExpirationRBTree.min();
+            const noExpiredOrdersLeft = this.expiration[nextOrderHashToExpire].greaterThan(
```
Comment (fabioberger) [inline packages/0x.js/src/order_watcher/expiration_watcher.ts]: Rename: `noExpiredOrdersLeft` to `hasNoExpiredOrders`

File: packages/0x.js/src/order_watcher/expiration_watcher.ts [hunk 20]
```diff
@@ -0,0 +1,76 @@
...
+                currentUnixTimestampMs.plus(this.expirationMarginMs),
+            );
+            const noActiveSubscription = _.isUndefined(this.orderExpirationCheckingIntervalIdIfExists);
```
Comment (fabioberger) [inline packages/0x.js/src/order_watcher/expiration_watcher.ts]: rename to `isSubscriptionActive`

**Pull request merged**